File: 66.PlusOne.go
package main


import (
	"fmt"
//	"reflect"
)

func plusOne(digits []int) []int {
	isJinwei := true
	if len(digits) == 1 && digits[0] == 0{
		digits[0] += 1
		return digits
	}
	for i := len(digits) - 1 ;i >= 0;i--{
		if isJinwei {
			if digits[i] == 9{
				isJinwei = true
				digits[i] = 0
			}else {
				isJinwei = false
				digits[i] += 1
			}
		}else{
			return digits
		}
	}
	if isJinwei {
		digits = append([]int{1}, digits...)
	}
	return digits
}


func main(){
	var a = []int{9, 9, 9}

	fmt.Println(plusOne(a))
}

File: Solutions/66.PlusOne.cpp
class Solution {
  public:
    vector<int> plusOne(vector<int> &digits) {
        int p = digits.size() - 1;
        if (digits[p] < 9) {
            digits[p]++;
            return digits;
        }
        bool carry = true;
        for (; p >= 0; p--) {
            if (carry) {
                if (digits[p] == 9) {
                    digits[p] = 0;
                    carry = true;
                } else {
                    digits[p]++;
                    carry = false;
                    break;
                }
            }
        }

        if (carry) {
            digits[0] = 1;
            digits.push_back(0);
        }
        return digits;
    }
};
